### PR TITLE
sysctl: Rename ignoreerrors to ignore_unknown_keys

### PR DESCRIPTION
**Pull request by winem**
This PR renames ignoreerrors to ignore_errors and keeps backward compatibility with ignorerrors. 

It's basically to be consistent with all other modules.

The sysctl module is the only one where the attribute is called ignoreerrors and not ignore_errors.
The new parameter is called ignore_errors but supports ignoreerrors as an alias.
The alias ensures that backward compatibility is given and this is not a breaking change.

##### SUMMARY
With this PR ignore_errors is allowed, too. I found it a bit unintuitive because it is not consistent with all other modules where the parameter is called ignore_errors with an underscore. The alias is to ensure backward compatibility because this cosmetically change should not break any playbooks out there. 

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
sysctl

##### ADDITIONAL INFORMATION
Documentation was updated and 2 tests to test both, ignoreerrors and ignore_errors were added.

### CODE REVIEW

**Comment by bcoca** [comment]
modules don't implement ignore_errors, it is a keyword of the task, this will lead to confusion, i almost prefer keeping the old name to disambiguate.

**Review by bcoca — CHANGES_REQUESTED** [review]
modules don't implement ignore_errors, it is a keyword of the task, this will lead to confusion, i almost prefer keeping the old name to disambiguate.

**Comment by winem** [comment]
@bcoca I totally agree here. It's a bit confusing especially when you combine both or use both in the same playbooks. To be honest, it even took me a moment to understand the different effects of ignoreerrors and ignore_errors. 

So I took another look at the other modules and it's really just the keyword that it's referenced there. 

I have mixed feelings about this PR and am not sure if consistency is worth the risk for confusion here. I let this up to you, guys. I also don't mind to change anything here. 

(Will take a look at the failing tests for Fedora, CentOS, etc. tomorrow. Already 3 a.m. here.)

**Comment by winem** [comment]
I thought a bit more about this and, if at all, it probably just makes sense to choose another name for the parameter (i.e. ignore_unknown_keys). ignore_errors on 2 different levels is something that should be avoided. 

The reason why I changed my opinion a bit is this from the previous comment: "So I took another look at the other modules and it's really just the keyword that it's referenced there." 

If there is no further input or request from anyone else with an opinion on this toic, I'll close this PR.

**Comment by winem** [comment]
@bcoca what do you think about the option to rename it to "ignore_unknown_keys"? I think that this is a small improvement and not as ambiguous as `ignoreerrors` (although the documentation is unambiguous here.
So this is more for people just reading a playbook without having all the modules and parameters in mind.

**Comment by gundalow** [comment]
`ignore_unknown_keys` sounds good

**Comment by winem** [comment]
Ok, will provide an updated PR.

**Review by bcoca — APPROVED** [review]

**Comment by mattclay** [comment]
Some of our test containers don't support reloading. Using `reload: no` as is done on the other tasks will work. Is there a reason why that can't be done for the new tests?

**Comment by bcoca** [comment]
@mattclay can we make it conditional? it would be nice to also test reloading .. since it is an integral part of the plugin

**Comment by winem** [comment]
I'm all in for conditionals on the conditional reload. Just let me know the OS or rather OS family (or any other condition) for containers that do not support reload. Any details are appreciated. @mattclay

**Comment by winem** [comment]
Ready for review and further input. @mattclay your input is highly appreciated. 

Btw, is there another way than a new push to the repo to trigger shippable again? The tests failed because the file system was read-only.

ready_for_review

**Comment by mattclay** [comment]
@winem Probably the best you can do is to call `sysctl` to set a parameter and see if it fails. If it does, then avoid reload during the test. I expect that should allow testing reload on our RHEL VMs, while skipping on most (if not all) of our Docker containers.

**Comment by winem** [comment]
@mattclay here is my Christmas gift for you. 

I added a check to see if the sysctl command is available and we have a conditional on the tasks that I suspect to fail if it's not available. Please feel free to test it. My first tests look good. 

This affects just the tests where `sysctl_set: true` was used or the `sysctl` command was executed via the shell module. Let me know if I missed anything or you see further ways to improve the tests.

**Comment by winem** [comment]
@mattclay here is my Christmas gift for you.

I added a check to see if the sysctl command is available and use `reload: yes` only when it's available.

File: lib/ansible/modules/system/sysctl.py [hunk 1]
```diff
@@ -241,13 +242,13 @@ def set_token_value(self, token, value):
             thiscmd = "%s %s=%s" % (self.sysctl_cmd, token, value)
         elif self.platform == 'freebsd':
             ignore_missing = ''
-            if self.args['ignoreerrors']:
+            if self.args['gnore_unknown_keys']:
```
Comment (bcoca) [inline lib/ansible/modules/system/sysctl.py]: missing an 'i'

**Review by bcoca — CHANGES_REQUESTED** [review]

**Comment by mattclay** [comment]
Integration tests are still trying to reload on systems which don't support it:

https://app.shippable.com/github/ansible/ansible/runs/101007/38/tests

File: lib/ansible/modules/system/sysctl.py (hunk 1 above)
Comment (winem) [reply lib/ansible/modules/system/sysctl.py]: @bcoca thanks, this was a mean one. Looks like a typical vi... 
Fixed will push it with the fix for systems not supporting sysctl.

**Comment by winem** [comment]
For some reason, the RedHat-Container return a "read-only filesystem" error (I did not notice that it was related immediately, so I'm sorry for the delay. I thought we're talking about two issues here.)

Now I fixed it and we handle 2 cases: 
1. if the sysctl command is not available at all (I even changed the check from `command -v sysctl` to `sysctl -w ...` when troubleshooting the issue. 
2. we do not run the ignore_unknown_keys unit tests on RedHat. 

Let me know if this is ok for you. To me it feels "ok-ish" but I'm not sure how satisfied I am with that solution. 
 
ready_for_review

File: test/integration/targets/sysctl/tasks/main.yml [hunk 2]
```diff
@@ -143,31 +168,80 @@
...
+## 
+
+- name: set value for fictional key with ignore_unknown_keys
+  sysctl: 
+    name: k.doesnotexist
+    value: 1 
+    ignore_unknown_keys: yes
+    reload: "{{ sysctl_cmd_supported }}"
+  register: sysctl_ignore_unknown_keys0
+  when: ansible_os_family != 'RedHat'
```
Comment (mattclay) [inline test/integration/targets/sysctl/tasks/main.yml:216]: This results in skipping the tests on the only Linux VM we use in CI, which makes me think this is simply covering up an issue. To make sure this is correct we need to understand why it's failing.

In the case of our Docker containers we know it's a permissions issue, but that shouldn't be the case on our RHEL virtual machines.

File: test/integration/targets/sysctl/tasks/main.yml [hunk 3]
```diff
@@ -29,6 +29,28 @@
     path: "{{ output_dir_test }}"
     state: directory
 
+- name: check if sysctl command is available
+  shell: sysctl -w net.ipv4.ip_forward=0
```
Comment (mattclay) [inline test/integration/targets/sysctl/tasks/main.yml:33]: I recommend a setting less likely to interfere with regular system operation:

```suggestion
  shell: sysctl -w debug.exception-trace=1
```

It would also be a good idea to query the existing value and use that for the set, since we don't actually need to change the setting, just verify that we have permissions to do so.

Also, the test should have `needs/root` added to the aliases file, since it requires root permissions. Since we're potentially changing sysctl values, `destructive` should also be added.

File: test/integration/targets/sysctl/tasks/main.yml [hunk 4]
```diff
@@ -125,15 +147,18 @@
     value: 1
     sysctl_set: yes
     reload: no
+  when: sysctl_cmd_supported
```
Comment (mattclay) [inline test/integration/targets/sysctl/tasks/main.yml:150]: Why is this needed? The existing tests pass without it and should only be updating `sysctl.conf` without writing the new values with `sysctl`.

File: test/integration/targets/sysctl/tasks/main.yml [hunk 5]
```diff
@@ -29,6 +29,28 @@
...
+  shell: sysctl -w net.ipv4.ip_forward=0
+  register: sysctl_command_test
+  ignore_errors: yes
+
+- debug:
+    var: sysctl_command_test
+    verbosity: 1
+
+- set_fact:
+    sysctl_cmd_supported: yes
```
Comment (mattclay) [inline test/integration/targets/sysctl/tasks/main.yml:42]: Wouldn't it be more accurate to call this `sysctl_write_supported`?

**Review by mattclay — CHANGES_REQUESTED** [review]

File: test/integration/targets/sysctl/tasks/main.yml (hunk 2 above)
Comment (winem) [reply test/integration/targets/sysctl/tasks/main.yml:216]: Totally agree and this is the part that does not really satisfy me. The test works fine on RedHat VMs and just fails on Docker Containers and (and that's the actual challenge I think) the response from the sysctl command in this case is not "key does not exist". Instead it was either read-only fs or something similiar. So the `sysctl -e` does not recognize it a invalid key and fails instead. 

I do not have an idea right now how to fix this or write a reliable and satisfying test here but I'm open for any input.

File: test/integration/targets/sysctl/tasks/main.yml (hunk 3 above)
Comment (winem) [reply test/integration/targets/sysctl/tasks/main.yml:33]: Oh, the debug keys are great for such a test. Will update it.

File: test/integration/targets/sysctl/tasks/main.yml (hunk 4 above)
Comment (winem) [reply test/integration/targets/sysctl/tasks/main.yml:150]: Here we test the sysctl_set function. This executes `sysclt -w` and fails on the docker containers, too. So we skip these tests if the first `sysctl -w` failed. This is again a bit "suboptimal" as it might fail for various reasons. If you have a better idea how to fix this, let me hear and learn. :)

File: test/integration/targets/sysctl/tasks/main.yml (hunk 5 above)
Comment (winem) [reply test/integration/targets/sysctl/tasks/main.yml:42]: Yes, now, when I review it, I got lost between the "Some of our test containers don't support reloading" and the file-system issues where I did not see the correlation right from the beginning. Will wait for your feedback on the other comments and am happy to change this then.

**Review by bcoca — APPROVED** [review]
